### PR TITLE
Refactor `ResXResourceWriter` to use `Dictionary<string, string>` instead of `Hashtable`

### PR DESCRIPTION
**Pull request by elachlan**
Refactor `ResXResourceWriter` to use `Dictionary<string, string>` instead of `Hashtable`.

Related: #8143

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8160)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs [hunk 1]
```diff
@@ -513,10 +510,8 @@ private void AddAssemblyRow(string elementName, string alias, string name)
 
         private string GetAliasFromName(AssemblyName assemblyName)
         {
-            _cachedAliases ??= new Hashtable();
-
-            string alias = (string)_cachedAliases[assemblyName.FullName];
-
+            _cachedAliases ??= new();
+            string alias = _cachedAliases[assemblyName.FullName];
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs:514]: wouldn't this result in key not found exception when `_cachedAliases` is assigned with `new` ?  Did you see a scenario where we are calling this with `_cachedAliases` being null?

Comment (elachlan) [reply src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs:514]: Ahh, I see. I will sort it. We will need to use TryGetValue.

File: src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs [hunk 2]
```diff
@@ -513,11 +510,8 @@ private void AddAssemblyRow(string elementName, string alias, string name)
 
         private string GetAliasFromName(AssemblyName assemblyName)
         {
-            _cachedAliases ??= new Hashtable();
-
-            string alias = (string)_cachedAliases[assemblyName.FullName];
-
-            if (string.IsNullOrEmpty(alias))
+            _cachedAliases ??= new();
+            if (!_cachedAliases.TryGetValue(assemblyName.FullName, out string alias))
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs:514]: I know this is existing code but it is flawed.  It was okay with `Hashtable` but replacing it with Dictionary prompts to refactor this code. Assigning an empty dictionary and immediately trying to access the index is not efficient. 
Is assignment necessary in this method? can we simply do   ` if (_cachedAliases == null ) return null ;`

Comment (elachlan) [reply src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs:514]: No. If `GetAliasFromName` is called, it needs to return an alias string. 

The logic is basically saying, create a cache if we haven't created one already. Check if the value exists in the cache, if it doesn't exist in the cache, add the alias to the cache and return said alias.

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs:514]: :). you are right. i read next line checking as non null.

**Review by dreddy-work — DISMISSED** [review]
LGTM.

File: src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs (hunk 2 above)
Comment (elachlan) [reply src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs:514]: Thank you for being a sanity check!

Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs:514]: This isn't exactly the same. Without null annotations turned on this is a dictionary of `<string?, string?>`. Please check that the returned value isn't null or empty as before.

**Review by JeremyKuhne — CHANGES_REQUESTED** [review]
See my comment on checking for null or empty.

File: src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs (hunk 2 above)
Comment (elachlan) [reply src/System.Windows.Forms/src/System/Resources/ResXResourceWriter.cs]: done.

**Review by RussKie — APPROVED** [review]

**Comment by elachlan** [comment]
@JeremyKuhne Could you please look over this one as well.

**Review by JeremyKuhne — APPROVED** [review]

**Pull request merged**